exodus_lambda/functions/verify_source.py

```diff
@@ -0,0 +1,64 @@
+import os
+import subprocess
+
+import boto3
+
+
+def lambda_handler(event, context):
+    """Verifies the commit delivered to the pipeline by GitHub has a
+    tag signed by a trusted entity.
+
+    Receives the commit ID from the pipeline via UserParameters.
+
+    This function must be manually deployed prior to pipeline creation.
+    """
+    # pylint: disable=unused-argument
+
+    job_id = event["CodePipeline.job"]["id"]
+    job_data = event["CodePipeline.job"]["data"]
+    git_commit = job_data["actionConfiguration"]["configuration"][
+        "UserParameters"
+    ]
+
+    bucket_name = "exodus-pipeline-artifacts"
+    whitelist_name = ".whitelist.gpg"
+
+    s3_client = boto3.client("s3")
+    pipeline_client = boto3.client("codepipeline")
+
+    try:
+        # 1) Download and import whitelisted GPG keys
+        s3_client.download_file(bucket_name, whitelist_name)
+        subprocess.run(
+            ["gpg", "--import", whitelist_name], check=True, text=True
+        )
+
+        # 2) Clone exodus-lambda and checkout deploy branch
+        repo_url = "https://github.com/release-engineering/exodus-lambda.git"
+        subprocess.run(
+            ["git", "clone", repo_url, "-b", "deploy"], check=True, text=True
+        )
+        os.chdir("exodus-lambda")
+
+        # 3) Retrieve tag for given commit
+        complete_process = subprocess.run(
+            ["git", "describe", "--exact-match", git_commit],
+            check=True,
+            text=True,
+            capture_output=True,
+        )
+        git_tag = complete_process.stdout
+
+        # 4) Verify tag
+        subprocess.run(["git", "verify-tag", git_tag], check=True, text=True)
+
+        # 5) Update whitelist
+        s3_client.upload_file(whitelist_name, bucket_name)
+
+        # Report success to pipeline
+        pipeline_client.put_job_success_result(jobId=job_id)
```

Comment:
This made me wonder: do we really need a special pipeline client object, why can't we just return a value or raise if there's a problem?

But according to https://docs.aws.amazon.com/codepipeline/latest/userguide/action-reference-Lambda.html this is indeed required: "As part of the implementation of the Lambda function, there must be a call to either the PutJobSuccessResult API or PutJobFailureResult API."

That's annoying, I wish they had made it so that the function could simply return a value to indicate success.  It's probably designed that way so you can post the result asynchronously after the lambda function has returned.

Reply:
Yeah, there are a lot of small inconveniences with AWS. The put job success call also allows provides a way to pass variables back to the pipeline, which could be useful in some cases.